### ISSUE TITLE
Support passing arbitrary options to underlying Docker commands

### ISSUE THREAD

**Issue by Rabadash8820** [aws/aws-sam-cli#3453]
### Describe your idea/feature/enhancement

I wish SAM CLI would support passing arbitrary arguments to the underlying Docker commands, primarily from `sam build` to `docker build` and from `sam local ...` to `docker run`. SAM CLI already has pretty great support for passing _environment_ info to the underlying Docker commands, via `--container-env-var`, `--container-env-var-file`, `--config-env`, `--env-vars`, and the like. However, there are instances where developers need to tweak the networking (e.g., for dev containers, which I believe led to `sam local ...`'s `--container-host` option), [volumes](https://github.com/aws/aws-sam-cli/issues/1896) (for local testing), and other properties of the underlying Docker containers. I feel like the most forward-compatible solution would just be to let developers pass arbitrary arguments to Docker, sort of like the `--docker-arg` option in the Serverless Framework's [`invoke local` command](https://www.serverless.com/framework/docs/providers/aws/cli-reference/invoke-local#options).

### Proposal

I think this could be similar to how options are [passed to npm scripts](https://docs.npmjs.com/cli/v7/commands/npm-run-script). Essentially, `sam build`, `sam local ...`, and other commands that talk to Docker would use `--` to "escape" options that should be passed directly to Docker and not parsed by SAM CLI. For example:

```sh
sam local start-api --port 3001 -- --mount 'type=volume,src=<VOLUME-NAME>,dst=<CONTAINER-PATH>'
```

In this example:
- The `--port` option is parsed by SAM CLI
- The `--mount` option and its argument are passed directly to Docker.
 
This is actually the use case that I've been trying to get working, as I want to test my functions locally by having them write to a shared volume on my localhost, rather than S3 (currently, I have to use [David Gray's hack](https://github.com/dacgray/sam-local-start-api-container-image-volume-mount) to make this work).

Alternatively, SAM CLI could take a `--docker-arg` option like Serverless Framework's [`invoke local` command](https://www.serverless.com/framework/docs/providers/aws/cli-reference/invoke-local#options), where the Docker options become arguments to _that_ option:

```sh
sam local start-api --port 3001 --docker-arg '-v /var:/host_var'
```

This would make it a little easier for developers coming from Serverless Framework, and arguably make the serverless computing ecosystem more uniform...though I personally like the brevity of `--` better.

Things to consider:

I don't think this will require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model)

SAM CLI could continue to have feature-specific flags/options like `--docker-network` and `--container-host`, both for backwards compatibility, and to make common scenarios more "first class citizens" of the CLI. But these would ultimately become just shorthands for the general option I think.

**Comment by moelasmar** [comment]
Thanks @Rabadash8820 for purposing this idea.
My understanding is your main use case is the you need to mount local directory to the Lambda Emulator running docker the same as [this issue](https://github.com/aws/aws-sam-cli/issues/1896)

What I try to understand is beside this what other use cases that can be solved by introducing the new command option `--docker-arg` and could not be solved by using the `--docker-network`, `--container-env-vars`, `--container-host`.  We need to make sure that the flexibility of adding different docker args will not break the main purpose of the local lambda emulator.

**Comment by moelasmar** [comment]
The idea of mounting local directories is also mentioned in this issue #2589

**Comment by Rabadash8820** [comment]
Correct, looking through #1896 and #2589 is actually what led me to this idea. I realized that I would've already been able to do what I wanted if SAM CLI just had some way of passing arbitrary options through to Docker. The fact that Serverless Framework supports this, while SAM CLI does not, seems like reason enough to add it: to maintain feature parity with the competition.

You say you can't think of any other reason that developers would need `--docker-arg`. Well that's just it, the option is for all the scenarios that you and I haven't thought of. Issues like #1896 and #2589 would never come up again if developers could just pass arguments to Docker. It's the ultimate "future proofing" against these Docker-specific issues. I think such an option fits well with what SAM CLI is anyway: a pretty thin wrapper around the other tools within it: CloudFormation, CDK, Docker, package managers, etc.

**Comment by dacgray** [comment]
@moelasmar 

> We need to make sure that the flexibility of adding different docker args will not break the main purpose of the local lambda emulator.

I get you, there are dangers in passing in docker args or mounting a volume.  But there are also dangers in SAM being too restrictive.  Sam would be a far more versatile tool if it were more flexible.  If a dev makes a stupid change that works locally but not on Lambda, that's on him, not on SAM.

Perhaps a docker arg blacklist/whitelist would be a safe compromise?

---

If this feature is on the table, I'd very much like to suggest a `func` addition as I mentioned in https://github.com/aws/aws-sam-cli/issues/2589

Would be very useful when developing an API that consists of more than one function.  Lends itself to the microservice model.

**Comment by moelasmar** [comment]
Thanks @Rabadash8820, @dacgray for your feedbacks.

**Comment by ygaudet** [comment]
Another example would be to pass --add-host to map a DocumentDB proper URL to the host IP, which would allow to connect to a DB via a local ssh tunnel.

**Comment by sema4-bguldiken** [comment]
@moelasmar
Just to provide an example, a specific case would be  "--add-host" parameter  as mentioned in the following medium page.

https://medium.com/@TimvanBaarsen/how-to-connect-to-the-docker-host-from-inside-a-docker-container-112b4c71bc66

On Linux, to be able  to resolve "host.docker.internal", "--add-host=host.docker.internal:host-gateway"  needs to be provided when docker container is executed
